src/server/debug/server/database_dump.go

```diff
@@ -0,0 +1,251 @@
+				// truncate the table that was being dumped.
+				return writeRowPageToJSON(rows, w, lastPage)
+			}
+			return errors.EnsureStack(err)
+		}
+		if len(rows) == 0 {
+			break
+		}
+		if err := writeRowPageToJSON(rows, w, middlePage); err != nil {
+			lastPageErr := writeRowPageToJSON(rows, w, lastPage)
+			return errors.New(fmt.Sprintf("error writing page: %s: "+
+				"error writing last page: %s", err.Error(), lastPageErr.Error()))
+		}
+	}
+	return writeRowPageToJSON(rows, w, lastPage)
+}
+
+func (s *debugServer) getFirstPage(ctx context.Context, column string, pageSize int, table *pachsql.SchemaTable) ([]pachsql.RowMap, error) {
+	query := fmt.Sprintf(`
+		SELECT * FROM %s.%s
+		ORDER BY %s 
+		FETCH FIRST %d ROWS ONLY;
+	`, table.SchemaName, table.TableName, column, pageSize)
+	pageRows, err := pachsql.GetScannedRows(ctx, s.database, query)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	return pageRows, nil
+}
+
+func setToStringBasedOnType(item interface{}) func(item interface{}) string {
+	switch fmt.Sprint(reflect.TypeOf(item)) {
+	case "time.Time":
+		return func(item interface{}) string {
+			asTime, ok := item.(time.Time)
+			if !ok {
+				return ""
+			}
+			formattedTime := asTime.Format("2006-01-2 15:04:05.000000 -07:00")
+			return fmt.Sprintf("'%v'", formattedTime)
+		}
+	case "[]uint8":
+		return func(item interface{}) string {
+			asByteArray, ok := item.([]uint8)
+			if !ok {
+				return ""
+			}
+			asString := ""
+			for _, num := range asByteArray {
+				asString += fmt.Sprintf("%02X", num)
+			}
+			return fmt.Sprintf("byteain('\\x%s')", asString)
+		}
+	default:
+		return func(item interface{}) string {
+			return fmt.Sprintf("'%v'", item)
+		}
+	}
+}
+
+func (s *debugServer) writeQueryResponseToJSON(ctx context.Context, query string, w io.Writer) error {
+	scannedRows, err := pachsql.GetScannedRows(ctx, s.database, query)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	return writeAllRowsToJSON(scannedRows, w)
+}
+
+func writeAllRowsToJSON(rowMap []pachsql.RowMap, w io.Writer) error {
+	return writeRowsToJSON(rowMap, w, "[{", "[\n\t{", "}]", "}\n]\n", "},", "},\n\t")
+}
+
+func writeRowPageToJSON(rowMap []pachsql.RowMap, w io.Writer, p page) error {
+	var replacerOpts []string
+	switch p {
+	case firstPage:
+		replacerOpts = []string{"[{", "[\n\t{", "}]", "}", "},", "},\n\t"}
+	case middlePage:
+		replacerOpts = []string{"[{", ",\n\t{", "}]", "}", "},", "},\n\t"}
+	case lastPage: // in this case, just write the end of the dictionary
+		if _, err := w.Write([]byte("\n]")); err != nil {
+			return errors.EnsureStack(err)
+		}
+		return nil
+	default:
+		return errors.New("page should be either firstPage, middlePage, or lastPage")
+	}
+	return writeRowsToJSON(rowMap, w, replacerOpts...)
+}
+
+func writeRowsToJSON(rowMap []pachsql.RowMap, w io.Writer, opts ...string) error {
+	marshalledRows, err := json.Marshal(rowMap)
```

Comment:
Sure, I can try that out and see if that simplifies things. We'll have slightly more overhead since we are calling marshal() more, but that probably won't be the bottleneck here

Reply:
I did this refactor and while it is 'about' as complicated, it is definitely more readable.